src/overlay/Peer.cpp

```diff
@@ -699,102 +794,168 @@ Peer::sendMessage(std::shared_ptr<StellarMessage const> msg, bool log)
 void
 Peer::sendAuthenticatedMessage(std::shared_ptr<StellarMessage const> msg)
 {
-    AuthenticatedMessage amsg;
-    amsg.v0().message = *msg;
-    if (msg->type() != HELLO && msg->type() != ERROR_MSG)
-    {
-        ZoneNamedN(hmacZone, "message HMAC", true);
-        amsg.v0().sequence = mSendMacSeq;
-        amsg.v0().mac =
-            hmacSha256(mSendMacKey, xdr::xdr_to_opaque(mSendMacSeq, *msg));
-        ++mSendMacSeq;
-    }
-    xdr::msg_ptr xdrBytes;
-    {
-        ZoneNamedN(xdrZone, "XDR serialize", true);
-        xdrBytes = xdr::xdr_to_msg(amsg);
-    }
-    this->sendMessage(std::move(xdrBytes));
+    auto cb = [msg](std::shared_ptr<Peer> self) {
+        {
+            // No need to hold the lock for the duration of this function:
+            // simply check if peer is shutting down, and if so, avoid putting
+            // more work onto the queues. If peer shuts down _after_ we already
+            // placed the message, any remaining messages will still go through
+            // before we close the socket, so this should be harmless.
+            RECURSIVE_LOCK_GUARD(self->mStateMutex, guard);
+            if (self->shouldAbort(guard))
+            {
+                return;
+            }
+        }
+
+        // Synchronous code block that executes TCPPeer::sendMessage preserves
+        // ordering of messages, which is important here, because we assign
+        // auth sequence numbers.
+        AuthenticatedMessage amsg;
+        amsg.v0().message = *msg;
+        if (msg->type() != HELLO && msg->type() != ERROR_MSG)
+        {
+            ZoneNamedN(hmacZone, "message HMAC", true);
+            amsg.v0().sequence = self->mSendMacSeq;
+            amsg.v0().mac =
+                hmacSha256(self->mSendMacKey,
+                           xdr::xdr_to_opaque(self->mSendMacSeq.load(), *msg));
+            self->mSendMacSeq++;
+        }
```

Comment:
This is perhaps a bit minor since I _think_ this is the only code path touching the MAC sequence number and I _think_ there's a level of protocol-handshake-enforced mutual exclusion between potential threads that can wind up here, but this access pattern looks like it's supposed to be logically atomic and as-written it's more of a race. Here's a sketch of a more-atomic usage:

```suggestion
            auto seq = self->mSendMacSeq.fetch_add(1) + 1;
            amsg.v0().sequence = seq;
            amsg.v0().mac =
                hmacSha256(self->mSendMacKey,
                           xdr::xdr_to_opaque(seq, *msg));
        }
```